File: Services/java-services/Core/CoreApplication/src/main/java/ai/omnicure/core/application/cache/ICacheService.java
package ai.omnicure.core.application.cache;

import java.time.Duration;
import java.util.Map;
import java.util.function.Supplier;

public interface ICacheService {

    void setString(String key, String value, Duration expiry);

    <T> void set(String key, T value, Duration expiry);

    String getString(String key);

    <T> T get(String key, Class<T> type);

    <T> T getOrSet(String key, Supplier<T> factory, Duration expiry, Class<T> type);

    void remove(String key);

    void removeByPrefix(String prefix);

    boolean acquireLock(String lockKey, String lockValue, Duration expiry);

    boolean releaseLock(String lockKey, String lockValue);

    void publish(String channel, String message);

    void subscribe(String channel, java.util.function.Consumer<String> handler);

    void hashSet(String key, Map<String, String> fields, Duration expiry);

    Map<String, String> hashGetAll(String key);

    String hashGet(String key, String field);

    void hashDelete(String key, String field);
}

File: Services/dotnet-services/Core/OmniCure.Core.Application/Interfaces/Caching/ICacheService.cs
namespace OmniCure.Core.Application.Interfaces.Caching;

public interface ICacheService
{
    Task SetStringAsync(string key, string value, TimeSpan expiry, CancellationToken ct = default);
    Task SetAsync<T>(string key, T value, TimeSpan expiry, CancellationToken ct = default);
    Task<string?> GetStringAsync(string key, CancellationToken ct = default);
    Task<T?> GetAsync<T>(string key, CancellationToken ct = default);
    Task<T> GetOrSetAsync<T>(string key, Func<CancellationToken, Task<T>> factory, TimeSpan expiry, CancellationToken ct = default);
    Task RemoveAsync(string key, CancellationToken ct = default);
    Task RemoveByPrefixAsync(string prefix, CancellationToken ct = default);

    Task<bool> AcquireLockAsync(string lockKey, string lockValue, TimeSpan expiry, CancellationToken ct = default);
    Task<bool> ReleaseLockAsync(string lockKey, string lockValue, CancellationToken ct = default);

    Task PublishAsync(string channel, string message, CancellationToken ct = default);
    Task SubscribeAsync(string channel, Action<string> handler, CancellationToken ct = default);

    Task HashSetAsync(string key, Dictionary<string, string> fields, TimeSpan? expiry = null, CancellationToken ct = default);
    Task<Dictionary<string, string>> HashGetAllAsync(string key, CancellationToken ct = default);
    Task<string?> HashGetAsync(string key, string field, CancellationToken ct = default);
    Task HashDeleteAsync(string key, string field, CancellationToken ct = default);
}
